### PR TITLE
Speed up `nearest_common_ancestor`.

### PR DESCRIPTION
**Pull request by nnethercote**
`nearest_common_ancestor` can be made faster.

Here are all the benchmarks where one of the measurements improved by at least 1%.
```
clap-rs-check
	avg: -4.5%	min: -8.8%	max: -0.3%
clap-rs
	avg: -2.6%	min: -4.5%	max: 0.5%
script-servo
	avg: -1.7%	min: -3.6%	max: 0.0%
regression-31157
	avg: -1.5%	min: -2.6%	max: -0.4%
hyper
	avg: -1.2%	min: -2.5%	max: -0.0%
piston-image
	avg: -1.6%	min: -2.5%	max: 0.1%
regex
	avg: -1.2%	min: -2.2%	max: 0.0%
issue-46449
	avg: -1.8%	min: -2.1%	max: -0.7%
crates.io
	avg: -1.2%	min: -2.1%	max: 0.0%
hyper-check
	avg: -1.0%	min: -2.1%	max: -0.1%
clap-rs-opt
	avg: -1.4%	min: -2.0%	max: -0.3%
piston-image-check
	avg: -1.2%	min: -1.9%	max: -0.1%
regex-check
	avg: -0.5%	min: -1.8%	max: -0.1%
syn
	avg: -1.1%	min: -1.7%	max: -0.1%
tokio-webpush-simple-check
	avg: -1.1%	min: -1.6%	max: -0.3%
tokio-webpush-simple
	avg: -1.2%	min: -1.6%	max: -0.0%
helloworld-check
	avg: -1.4%	min: -1.6%	max: -1.2%
deeply-nested
	avg: -1.2%	min: -1.4%	max: -0.8%
encoding-check
	avg: -0.8%	min: -1.3%	max: -0.3%
unify-linearly-check
	avg: -1.0%	min: -1.3%	max: -0.8%
script-servo-check
	avg: -0.6%	min: -1.3%	max: 0.0%
regression-31157-check
	avg: -0.9%	min: -1.2%	max: -0.7%
script-servo-opt
	avg: -0.5%	min: -1.2%	max: 0.1%
deeply-nested-check
	avg: -0.8%	min: -1.2%	max: -0.7%
encoding
	avg: -0.7%	min: -1.1%	max: -0.3%
issue-46449-check
	avg: -0.9%	min: -1.1%	max: -0.6%
parser-check
	avg: -0.9%	min: -1.1%	max: -0.8%
html5ever
	avg: -0.5%	min: -1.0%	max: -0.0%
```

### CODE REVIEW

File: src/librustc/middle/region.rs [hunk 1]
```diff
@@ -706,7 +706,7 @@ impl<'tcx> ScopeTree {
             // "Modeling closures" section of the README in
             // infer::region_constraints for more details.
             let a_root_scope = a_ancestors[a_index];
-            let b_root_scope = a_ancestors[a_index];
+            let b_root_scope = b_ancestors[b_index];
```
Comment (michaelwoerister) [inline src/librustc/middle/region.rs]: Nice catch!

**Comment by michaelwoerister** [comment]
Thanks, @nnethercote! I think @nikomatsakis had some ideas on how to exploit some domain knowledge to speed this. He probably should take a look at this.

r? @nikomatsakis

**Comment by nnethercote** [comment]
Here are things I've found from my measurements:
- Scope chains can be long, usually at least dozens of entries, and often 100+ entries.
- The vast majority of the time, the the nearest common ancestor is *much* closer to the front of the scope chain than the back. I.e. the two scope chains have a handful of elements that differ at the front and then a long common tail. This is why the new algorithm is a much better fit.

I'm not satisfied with the representation of `parent_map`, which is a HashMap mapping child scopes to parent scopes. It's hot and it feels to me like we should be able to use a different representation that doesn't require a hash table lookup for every child-to-parent lookup. I have some half-baked ideas that I am investigating, but they might be more suitable for a follow-up PR. Even if those ideas work out, I'm confident that this new algorithm will continue to be a better fit than the old one.

**Comment by nnethercote** [comment]
> I'm not satisfied with the representation of parent_map, which is a HashMap mapping child scopes to parent scopes. It's hot and it feels to me like we should be able to use a different representation that doesn't require a hash table lookup for every child-to-parent lookup. I have some half-baked ideas that I am investigating, but they might be more suitable for a follow-up PR.

I've done some work on this, and I haven't yet managed to make things any faster with the new representation. So I think it's definitely worth evaluating this PR as is.

**Review by nikomatsakis — APPROVED** [review]
This looks like the algorithm I wanted to do too -- thanks @nnethercote !

**Comment by nnethercote** [comment]
This looks like an infrastructure failure. Can someone please retry?

**Comment by nagbot-rs** [comment]
@bors: retry

* crates.io was down

**Comment by kirillkh** [comment]
Two remarks :
1. In the worst case, the new algorithm is quadratic in the combined length of the two stacks (`seen`). Did you try to use a hash table instead of SmallVec? 
2. It is possible to ~halve the number of comparisons being done in each iteration of the outer loop. Maintain two `seen` vecs instead of one. Add items from stack a into `seena` and items from stack b into `seenb`. Look for items from stack a in `seenb` and vice versa.

**Comment by nnethercote** [comment]
I did try a hash table. SmallVec was slightly faster for the rustc-perf suite.

Two Vecs would halve the number of comparisons, but it would require an extra heap allocation. I don't know which effect would be greater.

**Comment by kirillkh** [comment]
And actually I expect 2 SmallVecs to be a win, considering they are stack-allocated most of the time.

**Comment by nnethercote** [comment]
I filed #50365 for the two vectors change. Thank you for the suggestion, @kirillkh.

**Pull request merged**